junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,525 @@
+    extends plugins.PluginComponent with transform.Transform {
+
+    protected def newTransformer(
+        unit: global.CompilationUnit): global.Transformer = {
+      new ScalaJSJUnitPluginTransformer
+    }
+
+    val global: Global = ScalaJSJUnitPlugin.this.global
+    val phaseName: String = "junit-inject"
+    val runsAfter: List[String] = List("mixin")
+    override val runsBefore: List[String] = List("jscode")
+
+    class ScalaJSJUnitPluginTransformer extends global.Transformer {
+
+      import global._
+      import rootMirror.getRequiredClass
+
+      private val testAnnotationType =
+        getRequiredClass("org.junit.Test").toType
+
+      private val fixMethodOrderAnnotationType =
+        getRequiredClass("org.junit.FixMethodOrder").toType
+
+      private val annotationWhiteList = List(
+        testAnnotationType,
+        getRequiredClass("org.junit.Before").toType,
+        getRequiredClass("org.junit.After").toType,
+        getRequiredClass("org.junit.BeforeClass").toType,
+        getRequiredClass("org.junit.AfterClass").toType,
+        getRequiredClass("org.junit.Ignore").toType
+      )
+
+      private val scalaJSJUnitTestType =
+        getRequiredClass("org.scalajs.junit.ScalaJSJUnitTest").toType
+
+      private val jUnitClassMetadataType =
+        getRequiredClass("org.scalajs.junit.JUnitClassMetadata").toType
+
+      private val jUnitTestMetadataType =
+        getRequiredClass("org.scalajs.junit.JUnitTestMetadata").toType
+
+      private val jUnitMethodMetadataTypeTree =
+        TypeTree(getRequiredClass("org.scalajs.junit.JUnitMethodMetadata").toType)
+
+      type ScalaJSPlugin = NscPlugin {
+        /* Hack to be able to access `registerModuleExports` of ScalaJSPlugin
+         * without modifying it.
+         */
+        def PrepInteropComponent: AnyRef {
+          def genExportMember(sym: Symbol): List[Tree]
+          def registerModuleExports(sym: Symbol): Unit
+          def genNamedExport(defSym: Symbol, jsName: String, pos: Position): Unit
+        }
+      }
+
+      lazy val scalaJSPlugin = {
+        global.plugins.collectFirst {
+          case pl if pl.getClass.getName == "org.scalajs.core.compiler.ScalaJSPlugin" => pl
+        }.get.asInstanceOf[ScalaJSPlugin]
+      }
+
+      override def transform(tree: Tree): Tree = tree match {
+        case tree: PackageDef =>
+          def isClassWithJUnitAnnotation(tree: Tree): Boolean = tree match {
+            case clazz: ClassDef =>
+              clazz.impl.body.exists {
+                case ddef: DefDef => hasAnnotation(ddef, testAnnotationType)
+                case _            => false
+              }
+
+            case _ => false
+          }
+          val newStats = tree.stats.groupBy { // Group the class with its module
+            case clDef: ClassDef => Some(clDef.name)
+            case _               => None
+          }.iterator.flatMap {
+            case (Some(_), xs) if xs.exists(isClassWithJUnitAnnotation) =>
```

Comment:
How does this handle inherited tests? What's the story on them anyway? I know trait mixins work (by testing, see jsEnv tests)